### PR TITLE
KoalaBear, a possible alternative to BabyBear

### PR DESCRIPTION
**Pull request by SyxtonPrime**
One of the drawbacks of BabyBear which arises when working with arithmetization-oriented hash functions (e.g.  Poseidon/Poseidon2 and/or Rescue) is that order of the multiplicative group for the BabyBear field is divisible by both 3 and 5. This forces us to use the power map x^7 which is noticeably slower than x^3 and x^5 involving several more multiplications.

This PR introduces a new 31-bit field called KoalaBear which is the prime field P_KB = 2^31 - 2^24 + 1. (Note the similar structure to the BabyBear prime, P_BB = 2^31 - 2^27 + 1). Similarly to how P_BB is the 31-bit prime with the maximal 2 adicity, P_KB a 31-bit prime with maximal two adicity if we additionally demand that x^3 is an automorphism of the multiplicative group. Due to the similarity of P_KB and P_BB, the majority of the code in this PR is literally copied from the BabyBear repository with just a few constants needing to be changed.

In terms of speed improvements for Poseidon2, there is no change for width 16 as the gain in speed of moving to x^3 or x^5 is offset by a large increase to the number of partial rounds needed. However, for width 24, poseidon2 gets around 10% faster.

By random coincidence, the inverse operation is also slightly faster in KoalaBear taking 36 multiplication/square operations instead of 37 in BabyBear. This may be an indicator that it is possible to save an operation in the addition chain for BabyBear but I haven't been able to find such a saving yet.

As a final comment, we should mention that the KoalaBear prime is not unique and so other options could be investigated. In particular 2^30 + 2^27 + 2^24 + 1 has identical properties being both 31-bit and with multiplicative group not divisible by 3 or 5. Another interesting option would be 2^29 - 2^26 + 1 which has higher 2-adicity but is not 31-bit which might increase the complexity of addition and reduce security.

### CODE REVIEW

File: koala-bear/src/koala_bear.rs [hunk 1]
```diff
@@ -0,0 +1,578 @@
...
+        }
+        output[i].value = to_monty(input[i]);
+        i += 1;
+    }
+    output
+}
+
+#[inline]
+#[must_use]
+fn to_monty_64(x: u64) -> u32 {
```
Comment (unzvfu) [inline koala-bear/src/koala_bear.rs:449]: Nit: Might as well be `const` as in BabyBear.

File: koala-bear/src/koala_bear.rs [hunk 2]
```diff
@@ -0,0 +1,578 @@
...
+    (((x as u128) << MONTY_BITS) % P as u128) as u32
+}
+
+#[inline]
+#[must_use]
+fn from_monty(x: u32) -> u32 {
```
Comment (unzvfu) [inline koala-bear/src/koala_bear.rs:455]: Nit: Might as well be `const` as in BabyBear.

File: koala-bear/src/koala_bear.rs [hunk 3]
```diff
@@ -0,0 +1,578 @@
...
+    monty_reduce(x as u64)
+}
+
+/// Montgomery reduction of a value in `0..P << MONTY_BITS`.
+#[inline]
+#[must_use]
+pub(crate) fn monty_reduce(x: u64) -> u32 {
```
Comment (unzvfu) [inline koala-bear/src/koala_bear.rs:462]: Nit: Might as well be `const` as in BabyBear.

File: koala-bear/src/aarch64_neon/packing.rs [hunk 4]
```diff
@@ -0,0 +1,1085 @@
...
+        let t = aarch64::vaddq_u32(lhs, rhs);
+        let u = aarch64::vsubq_u32(t, P);
+        aarch64::vminq_u32(t, u)
+    }
+}
+
+// MONTGOMERY MULTIPLICATION
+//   This implementation is based on [1] but with changes. The reduction is as follows:
+//
+// Constants: P = 2^31 - 2^27 + 1
```
Comment (unzvfu) [inline koala-bear/src/aarch64_neon/packing.rs:176]: Nit: Update for KB.

File: koala-bear/src/aarch64_neon/packing.rs
Comment (unzvfu) [inline koala-bear/src/aarch64_neon/packing.rs:1]: This file is an old version of the corresponding BabyBear code. Is there any reason why your using that and not the newest version?

**Review by unzvfu — CHANGES_REQUESTED** [review]
Ok, looks good* over all. Apart from a few nits, just need to know why you haven't used Jacqui's most recent code for `aarch64_neon`.

*This extreme amount of code duplication is obviously highly undesirable, so could you please log an issue to the effect that a serious refactoring effort is required for BabyBear and KoalaBear.

File: koala-bear/src/aarch64_neon/packing.rs
Comment (SyxtonPrime) [reply koala-bear/src/aarch64_neon/packing.rs:1]: No, I just didn't realize it had been updated. (Clear downsides of code duplication...) I'll update it to use the modern code.

**Pull request merged**